### PR TITLE
allocation profiler: get stacks for pool allocs via wrapper function

### LINKED ISSUE #43688: allocs profiler: extend to record Types for all allocations - (fix `UnknownType` in allocations results)
The allocation profiler in https://github.com/JuliaLang/julia/pull/42768 doesn't provide _Types_ for all allocations. There are two types of codepaths which allocations go through:
1. generated code calling `jl_gc_alloc_(jl_ptls_t ptls, size_t sz, void *ty)` or `jl_gc_alloc_string` in `julia_internal.h`: these functions know the _type_ of the object being allocated.
    * which then (sometimes) calls `jl_gc_pool_alloc` / `jl_gc_big_alloc`
2. generated code calling `jl_gc_pool_alloc` or `jl_gc_big_alloc` in `gc.c` directly: these functions do _not_ know the type of the object being allocated.

The latter paths make it difficult to record the type of the allocation, since the type is not passed in — code is generated which calls it with just the size, and then sets the type later.

Another previous profiler PR, https://github.com/JuliaLang/julia/pull/33467, refactored the second codepath to take the type, and modifies the codegen to call the modified codepath, but hasn't been finished. For now, julia's Profile.Allocs functionality (introduced in https://github.com/JuliaLang/julia/pull/42768, https://github.com/JuliaLang/julia/pull/43868, https://github.com/JuliaLang/julia/pull/44043) doesn't do this.

Instead, for those types, we emit the type as `Profile.Allocs.UnknownType` in the allocations profile.

Experiments on a large codebase seem to show around ~40% of allocations have UnknownType, so the tool was still mostly useful. Still, getting the rest would be great!

-------

Fixing this would require either passing the type of the allocation through to those functions by changing the codegen, or removing the instrumentation from those functions and instead emitting the instrumentation directly in the codegen.

### PR DESCRIPTION
**Pull request by vilterp**
followup to https://github.com/JuliaLang/julia/pull/42768
fixes #43688 (gets the stacks, though not the types, of pool allocs)

Add a wrapper around jl_gc_pool_alloc to allow us to instrument it.

## Before

graphviz (1)

*: instrumented; red: currently missed

<details>
<summary>graphviz</summary>

```graphviz
digraph G {
  node [shape=box];

  generated_code [label="<generated code>"];

  new_array;
  jl_gc_alloc_string [label="jl_gc_alloc_string *"];
  jl_gc_alloc [label="jl_gc_alloc *"];
  jl_gc_managed_malloc [label="jl_managed_malloc *"]; // https://github.com/vilterp/julia/pull/20
  jl_gc_pool_alloc;
  jl_gc_big_alloc;

  {
    rank=same;
    jl_gc_pool_alloc;
    jl_gc_managed_malloc;
    jl_gc_big_alloc;
  }

  generated_code -> jl_gc_alloc_string;
  generated_code -> new_array;
  generated_code -> jl_gc_alloc;
  generated_code -> jl_gc_big_alloc [color=red];
  generated_code -> jl_gc_pool_alloc [color=red];

  jl_gc_alloc -> jl_gc_pool_alloc;
  jl_gc_alloc -> jl_gc_big_alloc;
  new_array -> jl_gc_alloc;
  new_array -> jl_gc_managed_malloc;
  jl_gc_alloc_string -> jl_gc_pool_alloc;
  jl_gc_alloc_string -> jl_gc_big_alloc;
}
```
</details>

## After

graphviz (2)

*: instrumented; red: currently missed; green: newly introduced function

<details>
<summary>graphviz</summary>

```graphviz
digraph G {
  node [shape=box];

  generated_code [label="<generated code>"];

  new_array;
  jl_gc_alloc_string [label="jl_gc_alloc_string *"];
  jl_gc_alloc [label="jl_gc_alloc *"];
  jl_gc_managed_malloc [label="jl_managed_malloc *"]; // https://github.com/vilterp/julia/pull/20
  jl_gc_pool_alloc [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc *"];
  jl_gc_pool_alloc_noinline [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc_noinline"];
  jl_gc_pool_alloc_inner;
  jl_gc_big_alloc;

  {
    rank=same;
    jl_gc_managed_malloc;
    jl_gc_pool_alloc_noinline;
    jl_gc_big_alloc;
  }

  generated_code -> jl_gc_alloc_string;
  generated_code -> new_array;
  generated_code -> jl_gc_alloc;
  generated_code -> jl_gc_big_alloc [color=red];
  generated_code -> jl_gc_pool_alloc;
  jl_gc_pool_alloc -> jl_gc_pool_alloc_inner  [style=dotted label="inlined"];
  jl_gc_pool_alloc_noinline -> jl_gc_pool_alloc_inner  [style=dotted  label="inlined"];


  jl_gc_alloc -> jl_gc_pool_alloc_noinline;
  jl_gc_alloc -> jl_gc_big_alloc;
  new_array -> jl_gc_alloc;
  new_array -> jl_gc_managed_malloc;
  jl_gc_alloc_string -> jl_gc_pool_alloc_noinline;
  jl_gc_alloc_string -> jl_gc_big_alloc;
}
```
</details>

Note that from the callers' perspectives, the code is unchanged. The only difference is that one call path now has an additional single (cold) branch, which can be enabled for profiling:
```c
    if (__unlikely(g_alloc_profile_enabled)) {
        _maybe_record_alloc_to_profile(val, size, jl_gc_unknown_type_tag);
    }
```

### CODE REVIEW

**Comment by vilterp** [comment]
Discussion from https://github.com/vilterp/julia/pull/17:
@vchuravy:
> I would prefer keeping the same name here, and renaming the inner internal implementation, which then also shouldn't be exported (and ideally inlined into this function).
> 
> Maybe call that one then jl_gc_pool_alloc_impl

@NHDaly:
> We talked about inlining it too, but we weren't sure about how to inline it _only_ into this function..
> 
> If we mark that one inline, it'll end up inlining into the normal callers of it, though.. So the _C_ functions (indicated in Pete's graph, above) that calls `jl_gc_pool_alloc` would now call `jl_gc_pool_alloc_impl` instead, and then it would be inlined.
> 
> So it'd become inlined into `jl_gc_alloc_` and `jl_gc_alloc_string`. Maybe that's okay?
> 
> Or would you want us to create yet a third separate wrapper function that gets called by those? So we would have three functions:
> - `jl_gc_pool_alloc_from_generated`: called from generated code
> - `jl_gc_pool_alloc`: called from `jl_gc_alloc_` and `jl_gc_alloc_string`
> - `jl_gc_pool_alloc_impl`: called from `jl_gc_pool_alloc`
> ? Or is that too weird? And the extra inlining is actually fine?

**Review by vtjnash — DISMISSED** [review]
Upon discussion, we don't think this is doable now, until we switch to OrcJIT v2 and have swappable function stub support available.

File: src/gc.c [hunk 1]
```diff
@@ -1196,6 +1196,18 @@ static NOINLINE jl_taggedvalue_t *add_page(jl_gc_pool_t *p) JL_NOTSAFEPOINT
     return fl;
 }
 
+// called into by LLVM-generated code
+JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc_outer(jl_ptls_t ptls, int pool_offset,
+                                          int osize)
+{
+    jl_value_t *val = jl_gc_pool_alloc(ptls, pool_offset, osize);
+    // zero out the type tag so it's not garbage
```
Comment (vchuravy) [inline src/gc.c:1204]: Why is this needed?

Comment (vilterp) [reply src/gc.c:1204]: we took it out, but I think something was lost in a rebase… fix coming shortly

Comment (NHDaly) [reply src/gc.c]: Yeah, repushed the commit; sorry about that. It's definitely not needed 👍

**Comment by vilterp** [comment]
Thanks for the info @vtjnash. I think I see how OrcJIT v2 to would make things more elegant. To make sure I understand:

* OrcJIT v2 will allow us to swap to an instrumented function without a wrapper, and without recompiling code?
* The approach in this PR seems bad because it introduces an additional function call to a hot codepath?

What's the expected timeline on the OrcJIT v2 migration? Is there an issue we can follow?

We'd like to merge something in the meantime, if possible… Perhaps we could get rid of the wrapper function either by
* adding a boolean parameter `instrument_alloc` to `jl_gc_pool_alloc` which is always `true` from generated code, and always `false` from internal usages, to avoid double counting when `jl_gc_pool_alloc` is called from paths that are already instrumented (`jl_gc_alloc`)
* giving up on recording types for the time being, and only instrumenting the leaves of the call graph above (i.e. `jl_gc_pool_alloc`, `jl_gc_managed_malloc`, and `jl_gc_big_alloc`)

Does either of these approaches seem viable? Thanks.

**Comment by vchuravy** [comment]
> OrcJIT v2 will allow us to swap to an instrumented function without a wrapper, and without recompiling code?

Yeah that would be to goal. It might even be not to hard to create a prototype for that, and would be needed for an alternative backend to the probe points in #43453.

> The approach in this PR seems bad because it introduces an additional function call to a hot codepath?

Yes, this is on the performance sensitive part of the allocator (and Jameson mentioned in offline discussion that we would like to inline the `pool_alloc` function fully). So any changes to that hot-path need to be performance neutral. Any additional call indirection is likely to be visible (hence my comment a while back that ideally we would need to callsite inline pool_alloc). 

> adding a boolean parameter instrument_alloc to jl_gc_pool_alloc

Might work, but I would like to see a performance study of that. i.e. we would like to avoid using an additional register xD

> giving up on recording types for the time being, and only instrumenting the leaves of the call graph above 

That could actually work, but is probably a lot less useful.

**Comment by vilterp** [comment]
Note from discussion w/ @JeffBezanson and @NHDaly:

We're going to try the last approach I proposed:
> giving up on recording types for the time being, and only instrumenting the leaves of the call graph above

In the meantime before OrcJIT v2, we'd prefer to have 100% of stacks, even without types, than missing 40% of allocations altogether. While types give a good overview of the workload, stacks give the essential info for drilling down and optimizing.

PR coming soon.

**Comment by vilterp** [comment]
Working on an allocations benchmark to try to capture perf overhead of this (and other approaches we might try). Initial results haven't shown a perf impact, although it may not be a sufficiently good benchmark: https://github.com/JuliaCI/BaseBenchmarks.jl/pull/291#issuecomment-1021574205

**Comment by vilterp** [comment]
Btw, this appears to be the spot in the Go memory allocator where they sample the stack trace if allocation profiling is enabled: https://github.com/golang/go/blob/c27a3592aec5f46ae18f7fd3d9ba18e69e2f1dcb/src/runtime/malloc.go#L1166-L1173

Nothing fancy (no compile-time flag), just an if statement (that gets branch predicted) like we're doing here.

**Comment by vilterp** [comment]
@nanosoldier `runbenchmarks("alloc", vs = ":master")`

**Comment by NHDaly** [comment]
I'm not sure if i have the permissions to kick off nanosoldier either... I haven't had great luck with it in the past. But I'll give it a shot:
@nanosoldier runbenchmarks("alloc", vs = ":master")

**Comment by NHDaly** [comment]
@nanosoldier `runbenchmarks("alloc", vs = ":master")`

**Comment by NHDaly** [comment]
Okay, i should have permissions now (thanks @vchuravy).

@nanosoldier `runbenchmarks("alloc", vs = ":master")`

**Comment by vchuravy** [comment]
I think @vtjnash needs to merge master into the nanosoldier branch.

**Comment by NHDaly** [comment]
@nanosoldier `runbenchmarks("alloc", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43868#issuecomment-1027293872) has completed, but no benchmarks were actually executed. Perhaps your tag predicate contains misspelled tags? cc @

**Comment by NHDaly** [comment]
trying again after the update the BaseBenchmarks (thanks @vtjnash!):

@nanosoldier `runbenchmarks("alloc", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43868#issuecomment-1027322517) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/0782a99_vs_b405562/report.md).

**Comment by NHDaly** [comment]
Okay! :) Nanosoldier has run and showed no regressions! In fact, it showed that the allocations benchmark was slightly faster with this PR, which is of course nonsense, but it means this change is in the noise, as we predicted.

Thanks for pushing for the benchmarking, that makes me feel much better knowing we've measured the impact carefully! 😊 

@vtjnash and @vchuravy can you give a final review now? Thanks very much!

File: src/gc.c [hunk 2]
```diff
@@ -1253,6 +1253,23 @@ JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc(jl_ptls_t ptls, int pool_offset,
     return jl_valueof(v);
 }
 
+// instrumented version of jl_gc_pool_alloc_inner called into by LLVM-generated code
```
Comment (NHDaly) [inline src/gc.c:1256]:
```suggestion
// Instrumented version of jl_gc_pool_alloc_inner, called into by LLVM-generated code.
```

**Review by NHDaly — APPROVED** [review]
LGTM. Thanks @vilterp for pushing this forward. This is a nice, clean implementation. 👍 
(I co-authored though, of course, so we still need another review.)

Thanks!

**Review by vchuravy — APPROVED** [review]

**Comment by NHDaly** [comment]
@vtjnash: We talked offline on slack last week and you said you weren't sure if you'd be able to get to reviewing this.

I'm going to merge it now, since we've had reviews from Valentin and talked it through with you in-person! Please still feel free to leave any review feedback you have, and we can always get to it in follow up PRs. ❤️ thanks again for your help! 😊
